docs/ops/state/checkpoints.md

```diff
@@ -99,3 +99,296 @@ above).
+
+* Then, the older checkpoint ``CP 1`` will be deleted because we have configured the number of retained checkpoints to
+two.
+
+* As part of the deletion, the counts for all files referenced by ``CP 1``, (``sstable-(1)`` and ``sstable-(2)``), is
+decreased by 1.
+
+Even though ``CP 1`` is logically deleted, we can see that all the files it created have still a reference count greater 
+0 and we cannot yet physically delete them from stable storage. Our graphic shows the counts after ``CP 3`` was
+registered and ``CP 1`` was deleted.
+
+For our last checkpoint ``CP 4``, RocksDB has merged ``sstable-(4)``, ``sstable-(5)``, and another ``sstable-(6)``,
+which was never observed at the time of a checkpoint, into ``sstable-(4,5,6)``. This file ``sstable-(4,5,6)`` is new for
+the checkpoint, and must be uploaded. We reference it together with ``sstable-(1,2,3)`` that was already known in
+``CP 4``. In the checkpoint coordinator’s shared state registry, the counts for ``sstable-(1,2,3)`` and
+``sstable-(4,5,6)`` are increased by 1. Then, ``CP 2`` is deleted as part of our retention policy. This decreases the
+counts for ``sstable-(1)``, ``sstable-(2)``, ``sstable-(3)``, and ``sstable-(4)`` by 1. This means that the counts for
+``sstable-(1)``, ``sstable-(2)``, and ``sstable-(3)`` have now dropped to 0 and they will be physically deleted from the
+stable storage. The final counts for ``CP 4`` after this step  are shown in the figure. This concludes our example for a
+sequence of 4 incremental checkpoints.
+
+#### Resolving Races for Concurrent Checkpoints
+
+We sometimes also have to resolve race conditions between concurrent checkpoints in incremental checkpointing. Flink can
+execute multiple checkpoints in parallel, and new checkpoints can start before previous checkpoints are confirmed as
+completed by the checkpoint coordinator to the backend. We need to consider this in our reasoning about which previous
+checkpoint can serve as a basis for a new incremental checkpoint. We are only allowed to reference shared state from a
+confirmed checkpoint, because otherwise we might attempt to reference a shared file that might still be deleted, e.g.
+when the assumed predecessor checkpoint still fails.
+
+This can lead to a situation were multiple checkpoints regard the same sstable files in RocksDB as new because no
+checkpoint that attempted to upload and register those sstable files has been confirmed, yet. To be on the safe side,
+checkpoints must always upload such files to stable storage independently, under unique names, until the sstable files
+have been registered by a completed checkpoint and the confirmation reached the backend. Otherwise, pending previous
+checkpoints might still fail, in which case their newly uploaded files are deleted, and future checkpoints would
+potentially attempt to reference deleted data.
+
+Sometimes, this upload policy will result in the same sstable file been uploaded more than once, from different
```

Comment:
This upload policy can result in the same sstable file being uploaded more than once, from different checkpoints.